switools/swisignature.py

```diff
@@ -144,9 +160,123 @@ def signSwiHandler( args ):
    rootCaFile = args.CAfile
    signatureFile = args.signature
    signingKeyFile = args.key
-   signSwi( swi, signingCertFile, rootCaFile, signatureFile, signingKeyFile )
+   signSwiAll( swi, signingCertFile, rootCaFile, signatureFile, signingKeyFile )
    print( 'SWI/X file %s successfully signed and verified.' % swi )
 
+def extractSignature( swi, destFile, sigFileName=None ):
+   sigFn = sigFileName
+   if not sigFn:
+      sigFn = signaturelib.SWI_SIG_FILE_NAME
+   ret = os.system( "set -e;"
+                    "destDir=$(readlink -f $(dirname %s));"
+                    "destFile=$(basename %s);"
+                    "swi=$(readlink -f %s);"
+                    "sigFile=%s;"
+                    "cd $destDir;"
+                    "unzip -o -q $swi $sigFile;"
+                    "mv $sigFile $destFile" % (
+                      destFile, destFile, swi, sigFn )
+                  )
+   if ret != 0:
+      print( "Error: Cannot extract signature file %s from  %s" % ( sigFn, swi ) )
+      sys.exit( SWI_SIGN_RESULT.CANNOT_EXTRACT_SIGNATURE )
+
+def getSignatureFile( swi, signatureFile ):
+   serviceName = 'swi-signing-service'
+   serviceBinary = shutil.which( serviceName )
+   if not serviceBinary:
+      print( "Error: signing service '%s' not found" % serviceName )
+      sys.exit( SWI_SIGN_RESULT.ERROR_SIGNING_SERVER_FAILED )
+
+   sha256 = prepareSwi( swi=swi, outfile=None, forceSign=True )
+   print( "%s sha256: %s" % ( os.path.basename( swi ), sha256 ) )
+   ret = subprocess.check_call( [ 'swi-signing-service', sha256, signatureFile ] )
+   if ret != 0:
+      print( "Error: signing-server '%s' failed" % serviceBinary )
+      sys.exit( SWI_SIGN_RESULT.ERROR_SIGNING_SERVER_FAILED )
+
+def signSwiAll( swi, signingCertFile, rootCaFile, signatureFile=None, signingKeyFile=None ):
+   # Sub-images ("optimizations") are extracted to /tmp. The utility 'swadapt' is
+   # handling that extraction. swadapt is found inside the image itself and is a
+   # statically linked i386 binary.
+   workDir = "/tmp/optims-%d" % os.getpid()
+   os.mkdir( workDir )
+   # Make sure the image we got is a swi file
+   if ( not os.path.isfile( swi ) or
+        os.system( "set -e; image=$(readlink -f %s); cd %s;"
+                   "unzip -o -q $image version" % ( swi, workDir ) ) ):
+      print( "Error: '%s' does not look like an EOS image" % swi )
+      shutil.rmtree( workDir )
+      sys.exit( SWI_SIGN_RESULT.ERROR_NOT_A_SWI )
+
+   try:
+      optims = signaturelib.getOptimizations( swi, workDir )
+      if optims is None or len( optims ) == 1 or "DEFAULT" in optims:
```

Comment:
With my suggested change to `signaturelib.getOptimizations`, this can become:
```suggestion
      optims = signaturelib.getOptimizations( zf )
      if len( optims ) <= 1 or "DEFAULT" in optims:
```